portal/src/graphql/portal/AuthenticationAuthenticatorSettings.tsx

```diff
@@ -188,18 +192,29 @@ const AuthenticationAuthenticatorSettings: React.FC<Props> = function Authentica
     },
   ];
 
-  const { primaryAuthenticators, secondaryAuthenticators } = constructListData(
-    effectiveAppConfig
-  );
+  const initialState = useMemo(() => {
+    return constructListData(effectiveAppConfig);
+  }, [effectiveAppConfig]);
 
   const [
     primaryAuthenticatorState,
     setPrimaryAuthenticatorState,
-  ] = React.useState(primaryAuthenticators);
+  ] = React.useState(initialState.primaryAuthenticators);
   const [
     secondaryAuthenticatorState,
     setSecondaryAuthenticatorState,
-  ] = React.useState(secondaryAuthenticators);
+  ] = React.useState(initialState.secondaryAuthenticators);
+
+  const [isFormModified, setIsFormModified] = useState(false);
+
+  useEffect(() => {
+    const screenState: AuthenticationAuthenticatorScreenState = {
+      primaryAuthenticators: primaryAuthenticatorState,
+      secondaryAuthenticators: secondaryAuthenticatorState,
+    };
+    const equalInitial = deepEqual(initialState, screenState, { strict: true });
+    setIsFormModified(!equalInitial);
+  }, [initialState, primaryAuthenticatorState, secondaryAuthenticatorState]);
```

Comment:
It seems that the state here is a bit complicated. Ideally we should have two copy of state. One for storing the initial state, one for storing the form state. The dirty flag should be derived instead of being a state.

```typescript
const initialState = useMemo(() => {
  return constructListData(effectiveAppConfig);
}, [effectiveAppConfig]);

const [formState, setFormState] = useState(initialState);

const isDirty = useMemo(() => {
  return !deepEqual(initialState, formState, { strict: true });
}, [initialState, formState]);
```